src/libasr/codegen/asr_to_wasm.cpp

```diff
@@ -1623,17 +1787,35 @@ class ASRToWASMVisitor : public ASR::BaseVisitor<ASRToWASMVisitor> {
     }
 
     void emit_string(std::string str) {
+        if (m_string_to_iov_loc_map.find(str) != m_string_to_iov_loc_map.end()) {
+            return;
+        }
+
         // Todo: Add a check here if there is memory available to store the
         // given string
-        wasm::emit_str_const(m_data_section, m_al, avail_mem_loc, str);
-        // Add string location in memory onto stack
+
         wasm::emit_i32_const(m_code_section, m_al, avail_mem_loc);
+        wasm::emit_i32_const(m_code_section, m_al, avail_mem_loc + 8);
+        wasm::emit_i32_store(m_code_section, m_al, wasm::mem_align::b8, 0);
+
+        wasm::emit_i32_const(m_code_section, m_al, avail_mem_loc + 4);
+        wasm::emit_i32_const(m_code_section, m_al, str.length());
+        wasm::emit_i32_store(m_code_section, m_al, wasm::mem_align::b8, 0);
+
+        m_string_to_iov_loc_map[str] = avail_mem_loc;
+
+        avail_mem_loc += 8;
+        wasm::emit_str_const(m_data_section, m_al, avail_mem_loc, str);
         avail_mem_loc += str.length();
         no_of_data_segments++;
+
+        avail_mem_loc = (int(ceil(avail_mem_loc / 4.0)) << 2); /* align mem location to 4 bytes */
```

Comment:
**Line 1812:** It seems that `fd_write` works only when its `iov` vector (it stores the start location of the string and the size of the string) is aligned by `4`. It doesn't seem to print any output when the location of the `iov` vector is not a multiple of `4`.

We previously had memory aligned by single byte. It seems we might need the memory alignment as `4` bytes in future.

Reply:
Good point.